### PR TITLE
DBZ-6259 Add missing description of various snapshot modes in the connector field definitions

### PR DESCRIPTION
**Pull request by ani-sha**
https://issues.redhat.com/browse/DBZ-6259

### CODE REVIEW

**Comment by ani-sha** [comment]
Hi @roldanbob, these descriptions are being sourced in RHOC UI. Could you please review the wordings? Thanks.

**Comment by roldanbob** [comment]
> Hi @roldanbob, these descriptions are being sourced in RHOC UI. Could you please review the wordings? Thanks.

@ani-sha Apologies for the delay. I've been busy completing updates for the downstream release, but I haven't forgotten you and will provide my suggestions soon.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java [hunk 1]
```diff
@@ -119,8 +119,11 @@ public class OracleConnectorConfig extends HistorizedRelationalDatabaseConnector
             .withGroup(Field.createGroupEntry(Field.Group.CONNECTOR_SNAPSHOT, 0))
             .withDescription("The criteria for running a snapshot upon startup of the connector. "
                     + "Options include: "
+                    + "'always' to specify that the connector run a snapshot each time it starts up; "
                     + "'initial' (the default) to specify the connector should run a snapshot only when no offsets are available for the logical server name; "
```
[the remark this answers is not in the document]
Comment (roldanbob) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java:123]: Same question as with the MySQL connector about `topic.prefix`.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 2]
```diff
@@ -264,6 +264,7 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
             .withDescription("The criteria for running a snapshot upon startup of the connector. "
                     + "Options include: "
                     + "'initial' (the default) to specify the connector should run a snapshot only when no offsets are available for the logical server name; "
```
[the remark this answers is not in the document]
Comment (roldanbob) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:266]: Same question as before about `topic.prefix`.

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 3]
```diff
@@ -264,6 +264,7 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
...
+                    + "'initial_only' same as 'initial' except the connector should stop after completing the snapshot and before it would normally read the binlog; "
                     + "'schema_only' to specify the connector should run a snapshot of the schema when no offsets are available for the logical server name. ");
```
[the remark this answers is not in the document]
Comment (roldanbob) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:268]: Same question as with the MySQL connector about `topic.prefix`.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 4]
```diff
@@ -760,6 +760,7 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
                     + "Options include: "
                     + "'when_needed' to specify that the connector run a snapshot upon startup whenever it deems it necessary; "
                     + "'schema_only' to only take a snapshot of the schema (table structures) but no actual data; "
+                    + "'schema_only_recovery' mode is used as recovery setting for a connector that has already been capturing changes, when you restart the connector, this setting enables recovery of a corrupted or lost database schema history topic; "
                     + "'initial' (the default) to specify the connector can run a snapshot only when no offsets are available for the logical server name; "
```
[the remark this answers is not in the document]
Comment (ani-sha) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java:764]: Yes @roldanbob, it should be reworded.

**Comment by ani-sha** [comment]
Thanks for the review @roldanbob! Could you also review the snapshot descriptions for `MongoDb` and `Postgres`?

**Comment by roldanbob** [comment]
> Thanks for the review @roldanbob! Could you also review the snapshot descriptions for `MongoDb` and `Postgres`?

Sure, I'd be happy to @ani-sha. It might take me a couple days to get to it. Would it be acceptable to have those review complete by sometime next week? I don't see those in this current PR, though. Do you want me to just review the following files and create a PR for any changes?

https://github.com/ani-sha/debezium/blob/main/debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorConfig.java

https://github.com/ani-sha/debezium/blob/main/debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 5]
```diff
@@ -262,9 +262,10 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
...
-                    + "Options include: "
-                    + "'initial' (the default) to specify the connector should run a snapshot only when no offsets are available for the logical server name; "
-                    + "'schema_only' to specify the connector should run a snapshot of the schema when no offsets are available for the logical server name. ");
+                    + "Select one of the following snapshot options: "
+                    + "'initial' (default): If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connectors begins to stream changes from the transaction log.; "
+                    + "'initial_only': The connector performs a snapshot as it does for the 'initial' option, but after completing the snapshot, it does not stream changes from the transaction log.; "
+                    + "'schema_only': If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures only the schema (table structures), but not any table data.");
```
Comment (Naros) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:268]: If we mention that for `initial`, the connector begins to stream changes from the transaction log, the same should be appended to the `schema_only` option for consistency and clarity.  I also believe we should mention that the connector stops after the snapshot finishes for `initial_only`.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java [hunk 6]
```diff
@@ -118,9 +118,12 @@ public class OracleConnectorConfig extends HistorizedRelationalDatabaseConnector
...
+                    + "'always': The connector runs a snapshot every time that it starts.; "
+                    + "'initial' (default):  If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables.  After the snapshot completes, the connector begins to stream changes from the binlog. "
+                    + "'initial_only': The connector performs a snapshot as it does for the  'initial' option, but after completing the snapshot, it does not stream changes from the redo logs.; "
+                    + "'schema_only': If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures only the schema (table structures), but not any table data.; "
+                    + "'schema_only_recovery': The connector performs a snapshot that captures only the database schema history. The connector then transitions to streaming from the redo logs. Use this setting to restore a corrupted or lost database schema history topic. Do not use if the database schema was modified after the connector stopped.");
```
Comment (Naros) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/OracleConnectorConfig.java:126]: As mentioned above, all options transition to streaming after their respective snapshot mode concludes, we may want to make that clear if we're doing so specifically for `initial`.  Furthermore, we should clarify the connector successfully stops with `initial_only` after the snapshot is performed.  And lastly, `initial` should reference "redo logs" and not "binlog".

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 7]
```diff
@@ -757,12 +757,13 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
...
+                    + "'schema_only_recovery': The connector performs a snapshot that captures only the database schema history. The connector then transitions back to streaming. Use this setting to restore a corrupted or lost database schema history topic. Do not use if the database schema was modified after the connector stopped.; "
+                    + "'initial' (default): If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connectors begins to stream changes from the binlog.; "
+                    + "'initial_only': The connector performs a snapshot as it does for the  'initial' option, but after completing the snapshot, it does not stream changes from the binlog.; "
+                    + "'never': The connector does not run a snapshot. Upon first startup, the connector immediately begins reading from the beginning of the binlog. "
```
Comment (Naros) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java:766]: Similar to the others, the wording about transitioning to streaming and capturing changes from the binlog apply to all these modes except `initial_only`, and yet we only clarify this for `initial_only`.  Is there a way we could reword this to be clear?

**Review by Naros — COMMENTED** [review]
Just a few things inline.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 8]
```diff
@@ -757,12 +757,13 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
...
-                    + "Options include: "
-                    + "'when_needed' to specify that the connector run a snapshot upon startup whenever it deems it necessary; "
-                    + "'schema_only' to only take a snapshot of the schema (table structures) but no actual data; "
-                    + "'initial' (the default) to specify the connector can run a snapshot only when no offsets are available for the logical server name; "
-                    + "'initial_only' same as 'initial' except the connector should stop after completing the snapshot and before it would normally read the binlog; and"
-                    + "'never' to specify the connector should never run a snapshot and that upon first startup the connector should read from the beginning of the binlog. "
+                    + "Select one of the following snapshot options: "
+                    + "'when_needed': On startup, the connector runs a snapshot if one is needed.; "
+                    + "'schema_only': If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures only the schema (table structures), but not any table data.; "
```
Comment (roldanbob) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java:762]:
```suggestion
                    + "'schema_only': If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures only the schema (table structures), but not any table data. After the snapshot completes, the connector begins to stream changes from the binlog.; "
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 9]
```diff
@@ -696,13 +696,13 @@ public static AutoCreateMode parse(String value, String defaultValue) {
...
             .withDescription("The criteria for running a snapshot upon startup of the connector. "
-                    + "Options include: "
-                    + "'always' to specify that the connector run a snapshot each time it starts up; "
-                    + "'initial' (the default) to specify the connector can run a snapshot only when no offsets are available for the logical server name; "
-                    + "'initial_only' same as 'initial' except the connector should stop after completing the snapshot and before it would normally start emitting changes;"
-                    + "'never' to specify the connector should never run a snapshot and that upon first startup the connector should read from the last position (LSN) recorded by the server; and"
-                    + "'exported' deprecated, use 'initial' instead; "
-                    + "'custom' to specify a custom class with 'snapshot.custom_class' which will be loaded and used to determine the snapshot, see docs for more details.");
+                    + "Select one of the following snapshot options: "
+                    + "'always': The connector runs a snapshot every time that it starts; "
```
Comment (roldanbob) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:700]:
```suggestion
                    + "'always': The connector runs a snapshot every time that it starts. After the snapshot completes, the connector begins to stream changes from the transaction log.; "
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 10]
```diff
@@ -696,13 +696,13 @@ public static AutoCreateMode parse(String value, String defaultValue) {
...
+                    + "'initial' (default):  If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connector begins to stream changes from the binlog. "
```
Comment (roldanbob) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:701]:
```suggestion
                    + "'initial' (default):  If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connector begins to stream changes from the transaction log. "
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 11]
```diff
@@ -696,13 +696,13 @@ public static AutoCreateMode parse(String value, String defaultValue) {
...
+                    + "'initial_only': The connector performs a snapshot as it does for the  'initial' option, but after completing the snapshot, it does not stream changes from the redo logs.; "
```
Comment (roldanbob) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:702]:
```suggestion
                    + "'initial_only': The connector performs a snapshot as it does for the  'initial' option, but after the connector completes the snapshot, it stops, and does not stream changes from the transaction log.; "
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 12]
```diff
@@ -696,13 +696,13 @@ public static AutoCreateMode parse(String value, String defaultValue) {
...
+                    + "'never': The connector does not run a snapshot. Upon first startup, the connector immediately begins reading from the beginning of the binlog. "
```
Comment (roldanbob) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:703]:
```suggestion
                    + "'never': The connector does not run a snapshot. Upon first startup, the connector immediately begins reading from the beginning of the transaction log. "
```

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java [hunk 13]
```diff
@@ -696,13 +696,13 @@ public static AutoCreateMode parse(String value, String defaultValue) {
...
+                    + "'exported': deprecated, use 'initial' instead; "
+                    + "'custom': to specify a custom class with 'snapshot.custom_class' which will be loaded and used to determine the snapshot, see docs for more details.");
```
Comment (roldanbob) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java:705]:
```suggestion
                    + "'exported': This option is deprecated; use 'initial' instead.; "
                    + "'custom': The connector loads a custom class  to specify how the connector performs snapshots. For more information, see Custom snapshotter SPI  in the PostgreSQL connector documentation.");
```

If it's possible to add a link, you could hyperlink **Custom snapshotter SPI** to https://debezium.io/documentation/reference/stable/connectors/postgresql.html#postgresql-custom-snapshot

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 14]
```diff
@@ -276,9 +276,10 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
...
```
Comment (roldanbob) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:282]:
```suggestion
                    + "'schema_only': If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures only the schema (table structures), but not any table data. After the snapshot completes, the connector begins to stream changes from the transaction log.");
```

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java [hunk 15]
```diff
@@ -276,9 +276,10 @@ public static SnapshotIsolationMode parse(String value, String defaultValue) {
             .withWidth(Width.SHORT)
             .withImportance(Importance.LOW)
             .withDescription("The criteria for running a snapshot upon startup of the connector. "
-                    + "Options include: "
-                    + "'initial' (the default) to specify the connector should run a snapshot only when no offsets are available for the logical server name; "
-                    + "'schema_only' to specify the connector should run a snapshot of the schema when no offsets are available for the logical server name. ");
+                    + "Select one of the following snapshot options: "
+                    + "'initial' (default): If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connectors begins to stream changes from the transaction log.; "
```
Comment (roldanbob) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerConnectorConfig.java:280]:
```suggestion
                    + "'initial' (default): If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connector begins to stream changes from the transaction log.; "
```

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 16]
```diff
@@ -757,12 +757,13 @@ public static SecureConnectionMode parse(String value, String defaultValue) {
...
+                    + "'schema_only_recovery': The connector performs a snapshot that captures only the database schema history. The connector then transitions back to streaming. Use this setting to restore a corrupted or lost database schema history topic. Do not use if the database schema was modified after the connector stopped.; "
+                    + "'initial' (default): If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connectors begins to stream changes from the binlog.; "
```
Comment (roldanbob) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java:764]:
```suggestion
                    + "'initial' (default): If the connector does not detect any offsets for the logical server name, it runs a snapshot that captures the current full state of the configured tables. After the snapshot completes, the connector begins to stream changes from the binlog.; "
```

**Review by roldanbob — CHANGES_REQUESTED** [review]
@ani-sha A few small edits for you. I think that I've got everything now, including updates to address the comments from @Naros  Thanks for your patience!

**Comment by ani-sha** [comment]
@roldanbob Thanks for the review. I have added all changes.

**Review by Naros — APPROVED** [review]

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java (hunk 13 above)
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java]: @Naros I agree on that one. But it is a low priority from my PoV.

Comment (Naros) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresConnectorConfig.java]: Added [DBZ-6378](https://issues.redhat.com/browse/DBZ-6378).

**Comment by Naros** [comment]
Applied to main (=2.3).
